### PR TITLE
Bugfix for missing calendar and increased memory limits

### PR DESCRIPTION
**Pull request by schucly**
Fixes missing calendar extension.
Fixes memory errors when added some plugins.

### CODE REVIEW

**Comment by tobybatch** [comment]
@kevinpapst I've got no problem with these but do we need the calendar extension and what is your advised mem limit?

I may move the memory to a run time arg one day

**Comment by kevinpapst** [comment]
Nope, calendar extension is not required. 
One paid plugin of a different developer uses one function from that extension, which is a problem, but not ours.
And there is nothing inside the calendar extension, which Kimai needs now or in the future, so we can leave it out.

It the development container, so it doesn't really matter. For normal dev work, 256M is already more than enough. The only task that eats up memory like crazy is PDF/Excel generation with large amount of data.

@schucly Posting a PR with "fixes" is great, but can you mention what you were trying to fix in the first place?

**Comment by clyso-dr** [comment]
In the prod container i have every second or third time on a kimai:reload a memory limit error, so a increase there to 256MB i would recommend. Also we are using the Controlling plugin and this ends up with an error because the function/extension cal_days_in_month is missing.

@kevinpapst It is may not needed for Kiami2 it self but I think your users are like to use plugins, so why we can not add this?

@tobybatch "I may move the memory to a run time arg one day" - I would appreciate that very much !

**Comment by clyso-dr** [comment]
Bildschirmfoto 2022-09-02 um 13 26 35

**Comment by clyso-dr** [comment]
Starting latest prod image on our instance with following plugins:

- AuditTrailBundle
- ControllingBundle
- EmptyDescriptionCheckerBundle
- HasWorkedBundle
- MetaFieldsBundle
- SickDayBundle
- VacationBundle

```
PHP Fatal error:  Allowed memory size of 134217728 bytes exhausted (tried to allocate 32768 bytes) in /opt/kimai/var/cache/de_/twig/d1/d19ae4c76a63ea2aa239dc8abb8abc07.php on line 404
[2022-10-19 06:11:43] php.CRITICAL: Fatal Error: Allowed memory size of 134217728 bytes exhausted (tried to allocate 32768 bytes) {"exception":"[object] (Symfony\\Component\\ErrorHandler\\Error\\OutOfMemoryError(code: 0): Error: Allowed memory size of 134217728 bytes exhausted (tried to allocate 32768 bytes) at /opt/kimai/var/cache/de_/twig/d1/d19ae4c76a63ea2aa239dc8abb8abc07.php:404)"} []
06:11:43 CRITICAL  [php] Fatal Error: Allowed memory size of 134217728 bytes exhausted (tried to allocate 32768 bytes) ["exception" => Symfony\Component\ErrorHandler\Error\OutOfMemoryError { …}]

In d19ae4c76a63ea2aa239dc8abb8abc07.php line 404:
                                                                               
  Error: Allowed memory size of 134217728 bytes exhausted (tried to allocate   
  32768 bytes)  
``` 

This message cames even two time before the container starts.

**Comment by kevinpapst** [comment]
Because the fault of the plugin developer does not need to be covered by the docker image. There is exactly ONE call to the `cal_days_in_month` function of the calendar extension, which can be easily changed to use DateTime objects - thus being compatible with core PHP.

Kimai clearly announces which extensions are required, calendar is not being one of them. 
The developer of the plugin (whom you are paying good money, compared to this repo here or Kimai core) probably missed that and accidentally used a function from the calendar extension. The better way to fix that is by opening bug reports for that paid plugin.

So I am strictly against adding that PHP extension in the official docker image. 
Because we would only be hiding the real problem.

The only time I saw an out of memory during cache reload was on a managed/shared hosting. Afaik 128M is the default setting on Ubuntu and Debian and I am running many systems that way without problems. 
Maybe that is a docker issue, which I cannot speak for.

I think 512M is way too much for a default setting. Especially as it seems to be the php.ini for both web and cli. 
You can pass in PHP settings to cli calls for PHP, maybe we should improve the default calls that are executed by the docker image. So what about for example:
```bash
php -d memory_limit=512M bin/console cache:clear
```

That should fix your issue as well, while keeping the default memory settings for all web processes. Having them as runtime argument would be great indeed ... @tobybatch are some of these calls in your scripts, that we should adjust as well?

File: Dockerfile [hunk 1]
```diff
@@ -261,7 +268,7 @@ RUN export COMPOSER_HOME=/composer && \
     composer --no-ansi require --working-dir=/opt/kimai laminas/laminas-ldap && \
     chown -R www-data:www-data /opt/kimai && \
     mkdir -p /opt/kimai/var/logs && chmod 777 /opt/kimai/var/logs && \
-    sed "s/128M/256M/g" /usr/local/etc/php/php.ini-development > /usr/local/etc/php/php.ini && \
+    sed "s/128M/512M/g" /usr/local/etc/php/php.ini-development > /usr/local/etc/php/php.ini && \
```
Comment (kevinpapst) [inline Dockerfile:271]: web process has nothing to do with memory issues on CLI / bash

File: Dockerfile [hunk 2]
```diff
@@ -281,6 +288,7 @@ RUN export COMPOSER_HOME=/composer && \
     composer --no-ansi clearcache && \
     composer --no-ansi require --working-dir=/opt/kimai laminas/laminas-ldap && \
     cp /usr/local/etc/php/php.ini-production /usr/local/etc/php/php.ini && \
+    sed "s/128M/512M/g" /usr/local/etc/php/php.ini-production > /usr/local/etc/php/php.ini && \
```
Comment (kevinpapst) [inline Dockerfile:291]: is there only one config for cli and web @tobybatch ?

**Comment by clyso-dr** [comment]
@kevinpapst Thanks for the clarifying, I already wrote the plugin developer again.

To the memory change please be aware of that we are using the container in Kubernetes so the change should not end up in a docker start call or something similar. :-)

**Comment by schucly** [comment]
Sorry for the request.
I'll close it.

**Comment by kevinpapst** [comment]
Raising such an issue is always good if we find spots to improve => so I am re-opening to clarify if we should update the cache clear (or other bon/console) calls.

**Comment by tobybatch** [comment]
I will move the mem limit to a run time arg. but we won't build in the calendar extension. The instruction for adding an extension are covered here: https://tobybatch.github.io/kimai2/build.html#extending-the-image

File: Dockerfile (hunk 2 above)
Comment (tobybatch) [reply Dockerfile:291]: There is a separate CLI php.ini. It is set to unlimited in the dev image. I'm going to unset that in the prod image as well.

https://github.com/tobybatch/kimai2/blob/main/Dockerfile#L265

**Comment by tobybatch** [comment]
@schucly & @clyso-dr I'm totally with Kevin about this. It's really good and helpful to raise issues like this. It has triggered me to move the memory override to a runtime arg.

https://github.com/tobybatch/kimai2/blob/main/assets/startup.sh#L40

https://github.com/tobybatch/kimai2/blob/main/assets/startup.sh#L40

File: Dockerfile (hunk 2 above)
Comment (kevinpapst) [reply Dockerfile:291]: Then we can skip my idea about adding `-d memory_limit=512M bin` to the bin/console calls. 
That is even better 👍

**Pull request closed without merging**